### PR TITLE
[JN-769] Persist events 

### PR DESCRIPTION
**Pull request by connorlbark**
#### DESCRIPTION

Persist events in the database when they occur. Show these events on the same page in the Admin UI that you can see notifications. Now, what used to be the "notifications" page is more like a "timeline" of all the granular things that has happened with an enrollee.

image

#### TO TEST:  *(simple manual steps for confirming core behavior -- used for pre-release checks)*

1. Restart database and fill out with `ourhealth` data
2. In admin ui, navigate to: https://localhost:3000/ourhealth/studies/ourheart/env/sandbox/participants/OHBASC/timeline
3. Ensure that the appropriate events are in the table and ordered properly in relation to notifications
4. Check postgres `event` table to make sure appropriate fields are filled out for each event

### CODE REVIEW

**Comment by devonbush** [comment]
This looks great, and works well when I test it locally.  I'll go through the code tomorrow, but a couple of meta-notes on the PR:
1. For UX changes, please always include a screenshot
2. for the steps, instead of providing a click sequence, just provide the url of where to go (unless the click sequence is important to see behavior).  So for this, you could just say "go to `https://localhost:3000/ourhealth/studies/ourheart/env/sandbox/participants/OHBASC/notifications`"

File: api-admin/src/main/resources/api/openapi.yml
Comment (devonbush) [inline api-admin/src/main/resources/api/openapi.yml:765]: thanks for linting this file

File: api-admin/src/main/resources/api/openapi.yml
Comment (devonbush) [inline api-admin/src/main/resources/api/openapi.yml]:
```suggestion
      summary: List of events relating to the specified enrollee
```

File: api-admin/src/main/resources/api/openapi.yml
Comment (connorlbark) [inline api-admin/src/main/resources/api/openapi.yml:159]: My IDE reformatted a bunch of stuff. They seem like good changes, but I'm happy to remove if we don't wanna include that. (For future PRs I reconfigured so it only reformats the lines I changed)

File: core/src/main/java/bio/terra/pearl/core/model/workflow/Event.java
Comment (devonbush) [inline core/src/main/java/bio/terra/pearl/core/model/workflow/Event.java]:
```suggestion
 * Simple record of an application event occurring, used primarily for debugging and customer support.
 * Note that these are records of application events (e.g. participant signup, kit received, etc...) as opposed to LogEvents, which log warnings and errors and is essentially just storing parts of the console log in the database.  
 * It might be the case that a single incident could trigger both an "Event" and a "LogEvent"
```

File: core/src/test/java/bio/terra/pearl/core/service/workflow/EventServiceTests.java
Comment (devonbush) [inline core/src/test/java/bio/terra/pearl/core/service/workflow/EventServiceTests.java]: Let's create a helper method for these assertions -- they seem to be repeated in all the test "confirmSavedEventMatches(...)"

File: ui-admin/src/study/participants/enrolleeView/EnrolleeTimeline.tsx
Comment (devonbush) [inline ui-admin/src/study/participants/enrolleeView/EnrolleeTimeline.tsx]: I think you can replace this function with lodash
`_startCase(_camelCase(str))`

File: ui-admin/src/study/participants/enrolleeView/EnrolleeTimeline.tsx
Comment (devonbush) [inline ui-admin/src/study/participants/enrolleeView/EnrolleeTimeline.tsx]: could this be declared instead as 
`type EventNotificationTableEntry = Notification | Event` ?

File: ui-admin/src/study/participants/enrolleeView/EnrolleeTimeline.tsx
Comment (devonbush) [inline ui-admin/src/study/participants/enrolleeView/EnrolleeTimeline.tsx]:
```suggestion
    const newTableData = [...(notifications as EventNotificationTableEntry[]), ...(events as EventNotificationTableEntry[])]
```
And if you change the type declaration of EventNotificationTableEntry to be Event | Notification you might not even need the `as` casting

File: ui-admin/src/study/participants/enrolleeView/EnrolleeTimeline.tsx
Comment (devonbush) [inline ui-admin/src/study/participants/enrolleeView/EnrolleeTimeline.tsx]: We now use tanstack react-table in our app wherever we have a table so that we don't have to rewrite sorting logic.  See MailingListView.tsx (visible athttps://localhost:3000/ourhealth/studies/ourheart/env/sandbox/mailingList) for the simplest example. Note how line 29 of that file sets the default initial sort, which is what you'll want to do to have this be sorted.

File: ui-admin/src/study/participants/enrolleeView/EnrolleeTimeline.tsx
Comment (devonbush) [inline ui-admin/src/study/participants/enrolleeView/EnrolleeTimeline.tsx]: yeah, if you use react-table, you can replace all this markup with `basicTableLayout(table)`

File: ui-admin/src/study/participants/enrolleeView/EnrolleeTimeline.tsx
Comment (devonbush) [inline ui-admin/src/study/participants/enrolleeView/EnrolleeTimeline.tsx]: IMO using the nullish coalesce operator is clearer at communicating "show this, but show blank if undefined"
```suggestion
                {row.deliveryType ?? ''}
```

**Review by devonbush — CHANGES_REQUESTED** [review]
Great stuff -- I'm pleased at how quickly this came together and how much potential there is here.  Some styling suggestions.  My main blocker is that we should update this to use the same table component we've now standardized on elsewhere in the admin tool.

File: ui-admin/src/study/participants/enrolleeView/EnrolleeTimeline.tsx
Comment (connorlbark) [reply ui-admin/src/study/participants/enrolleeView/EnrolleeTimeline.tsx]: Genius, I have much more JS experience than TS experience so stuff like this is actually really helpful to call out. Thanks!

File: ui-admin/src/study/participants/enrolleeView/EnrolleeTimeline.tsx
Comment (devonbush) [inline ui-admin/src/study/participants/enrolleeView/EnrolleeTimeline.tsx]: this will be cleaner if you use our `useLoadingEffect` hook.  See line 89 of MailingListView.tsx for an example -- that way the service calls will get wrapped in error handling, and we'll better communicate in code that this is loading the main data for the component to display.  It will also handle the isLoading logic for you

File: ui-admin/src/study/participants/enrolleeView/EnrolleeTimeline.tsx
Comment (devonbush) [inline ui-admin/src/study/participants/enrolleeView/EnrolleeTimeline.tsx]: Also, consider using Promise.all so the events and notifications are fetched in parallel.  Something like
```
const results = await Promise.all(
  Api.fetchNotifications(...),
  Api.fetchEnrolleeEvents(...)
)
```
Although performance is definitely not critical here, so if it's complicated or makes the code unwieldy, don't worry about it

File: ui-admin/src/study/participants/enrolleeView/EnrolleeTimeline.tsx
Comment (devonbush) [inline ui-admin/src/study/participants/enrolleeView/EnrolleeTimeline.tsx]: This feels weird to me to have a useEffect that's just combining the two arrays, it's essentially storing the same data in two different state variables.  

Let's eliminate the separate state variables for setEvents and setNotifications, and do this combination and calling of `setTableData` in the hook above this one

File: core/src/test/java/bio/terra/pearl/core/service/workflow/EventServiceTests.java
Comment (devonbush) [inline core/src/test/java/bio/terra/pearl/core/service/workflow/EventServiceTests.java:122]: nice!

File: core/src/main/java/bio/terra/pearl/core/service/workflow/EventService.java
Comment (devonbush) [inline core/src/main/java/bio/terra/pearl/core/service/workflow/EventService.java:129]: I don't think I mentioned this in the last review, but this was a great choice to make this method private -- that really clearly communicates that saving is handled internally

**Review by devonbush — CHANGES_REQUESTED** [review]
Looking good -- the table upgrade is a big win for readability.  One more pattern you should use that I missed the first time -- `useLoadingEffect`.  Thanks again for upgrading this component -- this was a very early component that was slapped together before we had a lot of patterns and utilities in place

File: api-admin/src/main/resources/api/openapi.yml
Comment (MatthewBemis) [reply api-admin/src/main/resources/api/openapi.yml:159]: Agreed, these are great formatting changes

File: core/src/main/java/bio/terra/pearl/core/service/workflow/EventService.java
Comment (connorlbark) [reply core/src/main/java/bio/terra/pearl/core/service/workflow/EventService.java:129]: If there are two things I got from my Northeastern education, it's to:
- Never make anything public
- Make so many tests you go dizzy from looking at them

**Review by MatthewBemis — APPROVED** [review]
Worked well when tested, code looks great too! Nice job!

File: ui-admin/src/study/participants/enrolleeView/EnrolleeTimeline.tsx
Comment (connorlbark) [reply ui-admin/src/study/participants/enrolleeView/EnrolleeTimeline.tsx]: This cleaned everything up a lot! thanks

File: ui-admin/src/study/participants/enrolleeView/EnrolleeTimeline.tsx
Comment (devonbush) [inline ui-admin/src/study/participants/enrolleeView/EnrolleeTimeline.tsx]: Let's not mix `await` and `then` -- you should use one or the other, (and generally prefer `await`).  Also, using `concat` in setTableData seems like it would duplicate entries if this component ever reloaded the list from the server, so I think it would be better (and cleaner), to just make a new array.
```suggestion
    const [notifications, events] = await Promise.all([
      Api.fetchEnrolleeNotifications(portal.shortcode, study.shortcode, currentEnv.environmentName, enrollee.shortcode),
      Api.fetchEnrolleeEvents(portal.shortcode, study.shortcode, currentEnv.environmentName, enrollee.shortcode)
    ])
    attachConfigsToNotifications(notifications)
    setTableData([...notifications, ...events])
   
  }, [enrollee.shortcode])
```

**Review by devonbush — APPROVED** [review]
One more change, and then this is good to go.  going ahead and approving now so you don't have to wait for my re-approval after the minor fixup.  thanks!

**Pull request merged**
